File: chapter3 - Stacks and Queues/3.3_stackOfPlates/JavaScript/SetOfStacks.js
const Stack = require("./Stack");

class SetOfStacks {
  constructor(capacity) {
    this.set = [];
    this.size = 0;
    this.vals = 0;
    this.capacity = capacity;
  };

  push(val) {
    if (!this.size || this.set[this.size - 1].isFull()) {
      const newStack = new Stack(this.capacity);
      newStack.push(val);
      this.set.push(newStack);
      this.size++;
    } else {
      const stack = this.set[this.size - 1];
      stack.push(val);
    }

    this.vals++;
  };

  pop() {
    if (this.size) {
      const stack = this.set[this.size - 1];
      const val = stack.pop();

      if (stack.isEmpty()) {
        this.set.pop();
        this.size--;
      }

      this.vals--;
      return val;
    }

    return undefined;
  };

  popAt(index) {
    if (this.size) {
      if (index < 0 || index >= this.vals) {
        this.set[this.set.length - 1].pop();
      } else {
        const stackIndex = this.findStackIndex(index);
        this.set[stackIndex].pop(index % this.capacity);
        this.shift(stackIndex);
      }

      this.vals--;
    }

    return undefined;
  };

  findStackIndex(index) {
    let count = 1;

    for (let x = 0; x < this.set.length; x++) {
      if (index < this.capacity * count) {
        return count - 1;
      }

      count++;
    }
  };

  shift(index) {
    if (index < this.set.length - 1) {
      const curr_stack = this.set[index];
      const next_stack = this.set[index + 1];
      

      while (!curr_stack.isFull() && !next_stack.isEmpty()) {
        const val = next_stack.stack.shift();
        curr_stack.push(val);
        next_stack.size--;
      }

      this.shift(index + 1);
    }

    if (this.set[index].isEmpty()) {
      this.set.pop();
      this.size--;
    }
  }
};

module.exports = SetOfStacks;

File: chapter3 - Stacks and Queues/3.3_stackOfPlates/Java_CTCI_Book/SetOfStacks.java
import java.util.ArrayList;

public class SetOfStacks {
  ArrayList<Stack> stacks = new ArrayList<Stack>();
  public int capacity;

  public SetOfStacks(int capacity) {
    this.capacity = capacity;
  }
  
  public void push(int v) {
    Stack last = this.getLastStack();

    if (last != null && !last.isFull()) { // add to last stack
      last.push(v);
    } else { // must create new Stack
      Stack stack = new Stack(capacity);
      
      stack.push(v);
      stacks.add(stack);
    }
  }

  public int pop() throws EmptyStackException {
    Stack last = this.getLastStack();
    if (last == null) throw new EmptyStackException();
    
    int v = last.pop();
    
    if (last.size == 0) this.stacks.remove(this.stacks.size() - 1);
    
    return v;
  }

  public boolean isEmpty() {
    Stack last = this.getLastStack();
    return last == null || last.isEmpty();
  }

  public Stack getLastStack() {
    if (this.stacks.size() == 0) return null;
    return this.stacks.get(this.stacks.size() - 1);
  }

  public int popAt(int index) {
    return this.leftShift(index, true);
  }

  public int leftShift(int index, boolean removeTop) {
    Stack stack = this.stacks.get(index);
    int removed_item;

    if (removeTop) removed_item = stack.pop();
    else removed_item = stack.removeBottom();

    if (stack.isEmpty()) {
      this.stacks.remove(index);
    } else if (this.stacks.size() > index + 1) {
      int v = leftShift(index + 1, false);
      stack.push(v);
    }

    return removed_item;
  }
}